metrics/cloud_images.py

```diff
@@ -102,118 +100,130 @@ def parse_simplestreams_for_images(cloud_name, image_type):
 
-    This returns the non-AWS-specific version of
-    parse_simplestreams_for_images' return tuples.
+    :param stats: a dict as described above
+    :param metrics: a list of influxdb-shaped metric dicts
     """
-    image_counts, aws_latest_serials = parse_simplestreams_for_images(
-        cloud_name, image_type)
-    latest_serials = {}
-    for release in aws_latest_serials:
-        # Some virt/storage combinations should be ignored, so we filter those
-        # out before we do anything else
-        aws_latest_serials[release] = {
-            k: v for k, v in aws_latest_serials[release].items()
-            if k not in AWS_VIRT_STORE_SKIPS.get(release, [])}
-        # aws_latest_serials contains an entry for each virt/storage combo; we
-        # want to use the oldest as the main cloud entry
-        latest_serials[release] = min(aws_latest_serials[release].values())
-        for virt_store in aws_latest_serials[release]:
-            aws_cloud_name = '{}:{}'.format(cloud_name, virt_store)
-            serial = aws_latest_serials[release][virt_store]
-            age = _determine_serial_age(serial)
-
-            tags = dict(image_type=image_type, cloud=aws_cloud_name,
-                        release=release)
-            metrics_collection.append(
-                _gen_influx_metric('current_serial', serial, **tags))
-            metrics_collection.append(
-                _gen_influx_metric('current_serial_age', age, **tags))
-
-    return image_counts, latest_serials
+    for image_type, clouds in stats.items():
+        for cloud_name, releases in clouds.items():
+            for release, stat_entry in releases.items():
+                _set_metrics_from_stat_item(metrics,
+                                            (image_type, cloud_name, release),
+                                            stat_entry)
+
+
+def _set_metrics_from_stat_item(metrics, path, stat_entry):
+    image_type, cloud_name, release = path
+
+    tags = dict(image_type=image_type, cloud=cloud_name, release=release)
+
+    for arch, stat in stat_entry['by-arch'].items():
+        metrics.append(_emit_metric(
+            'published',
+            stat['count'],
+            arch=arch,
+            **tags
+        ))
+
+    if 'latest_serial' in stat_entry:
+        metrics.append(_emit_metric(
+            'current_serial', stat_entry['latest_serial'], **tags))
+        metrics.append(_emit_metric(
+            'current_serial_age', stat_entry['age'], **tags))
+
+    if len(stat_entry['by-machine']) > 1:
+        for machine_type, stat in stat_entry['by-machine'].items():
+            if 'latest_serial' not in stat:
+                continue
+
+            tags['cloud'] = cloud_name + ':' + machine_type
+
+            metrics.append(_emit_metric(
+                'current_serial', stat['latest_serial'], **tags))
+            metrics.append(_emit_metric(
+                'current_serial_age', stat['age'], **tags))
```

Comment:
There is a subtle difference between what the original code did and this new code for AWS where we have this by-machine set of metrics. The old code would take the minimum of all of the by-machine metrics and use that as the latest value. This catches cases where the latest serial is only partially published and the grafana graphs end up showing an 'aws' serial that is older than we expect. This is currently the case for aws-govcloud where one of the by-machine entries did not get published for the most recent serial and leads to a difference in output. I have a diff here that resoves the difference: https://pastebin.canonical.com/p/kbSHyVWYwJ/.